File: ts-reference/examples/basic-2/generated-client/accounts/Counter.ts
import { PublicKey, Connection } from "@solana/web3.js"
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface CounterFields {
  authority: PublicKey
  count: BN
}

export interface CounterJSON {
  authority: string
  count: string
}

export class Counter {
  readonly authority: PublicKey
  readonly count: BN

  static readonly discriminator = Buffer.from([
    255, 176, 4, 245, 188, 253, 124, 25,
  ])

  static readonly layout = borsh.struct([
    borsh.publicKey("authority"),
    borsh.u64("count"),
  ])

  constructor(fields: CounterFields) {
    this.authority = fields.authority
    this.count = fields.count
  }

  static async fetch(
    c: Connection,
    address: PublicKey
  ): Promise<Counter | null> {
    const info = await c.getAccountInfo(address)

    if (info === null) {
      return null
    }
    if (!info.owner.equals(PROGRAM_ID)) {
      throw new Error("account doesn't belong to this program")
    }

    return this.decode(info.data)
  }

  static async fetchMultiple(
    c: Connection,
    addresses: PublicKey[]
  ): Promise<Array<Counter | null>> {
    const infos = await c.getMultipleAccountsInfo(addresses)

    return infos.map((info) => {
      if (info === null) {
        return null
      }
      if (!info.owner.equals(PROGRAM_ID)) {
        throw new Error("account doesn't belong to this program")
      }

      return this.decode(info.data)
    })
  }

  static decode(data: Buffer): Counter {
    if (!data.slice(0, 8).equals(Counter.discriminator)) {
      throw new Error("invalid account discriminator")
    }

    const dec = Counter.layout.decode(data.slice(8))

    return new Counter({
      authority: dec.authority,
      count: dec.count,
    })
  }

  toJSON(): CounterJSON {
    return {
      authority: this.authority.toString(),
      count: this.count.toString(),
    }
  }

  static fromJSON(obj: CounterJSON): Counter {
    return new Counter({
      authority: new PublicKey(obj.authority),
      count: new BN(obj.count),
    })
  }
}

File: examples/client-gen/basic_2/accounts/counter.py
import typing
from dataclasses import dataclass
from solders.pubkey import Pubkey
from solana.rpc.async_api import AsyncClient
from solana.rpc.commitment import Commitment
import borsh_construct as borsh
from anchorpy.coder.accounts import ACCOUNT_DISCRIMINATOR_SIZE
from anchorpy.error import AccountInvalidDiscriminator
from anchorpy.utils.rpc import get_multiple_accounts
from anchorpy.borsh_extension import BorshPubkey
from ..program_id import PROGRAM_ID


class CounterJSON(typing.TypedDict):
    authority: str
    count: int


@dataclass
class Counter:
    discriminator: typing.ClassVar = b"\xff\xb0\x04\xf5\xbc\xfd|\x19"
    layout: typing.ClassVar = borsh.CStruct(
        "authority" / BorshPubkey, "count" / borsh.U64
    )
    authority: Pubkey
    count: int

    @classmethod
    async def fetch(
        cls,
        conn: AsyncClient,
        address: Pubkey,
        commitment: typing.Optional[Commitment] = None,
        program_id: Pubkey = PROGRAM_ID,
    ) -> typing.Optional["Counter"]:
        resp = await conn.get_account_info(address, commitment=commitment)
        info = resp.value
        if info is None:
            return None
        if info.owner != program_id:
            raise ValueError("Account does not belong to this program")
        bytes_data = info.data
        return cls.decode(bytes_data)

    @classmethod
    async def fetch_multiple(
        cls,
        conn: AsyncClient,
        addresses: list[Pubkey],
        commitment: typing.Optional[Commitment] = None,
        program_id: Pubkey = PROGRAM_ID,
    ) -> typing.List[typing.Optional["Counter"]]:
        infos = await get_multiple_accounts(conn, addresses, commitment=commitment)
        res: typing.List[typing.Optional["Counter"]] = []
        for info in infos:
            if info is None:
                res.append(None)
                continue
            if info.account.owner != program_id:
                raise ValueError("Account does not belong to this program")
            res.append(cls.decode(info.account.data))
        return res

    @classmethod
    def decode(cls, data: bytes) -> "Counter":
        if data[:ACCOUNT_DISCRIMINATOR_SIZE] != cls.discriminator:
            raise AccountInvalidDiscriminator(
                "The discriminator for this account is invalid"
            )
        dec = Counter.layout.parse(data[ACCOUNT_DISCRIMINATOR_SIZE:])
        return cls(
            authority=dec.authority,
            count=dec.count,
        )

    def to_json(self) -> CounterJSON:
        return {
            "authority": str(self.authority),
            "count": self.count,
        }

    @classmethod
    def from_json(cls, obj: CounterJSON) -> "Counter":
        return cls(
            authority=Pubkey.from_string(obj["authority"]),
            count=obj["count"],
        )
